File: modules/rust_workloads/src/retained_growth.rs
//! Steadily grows a retained, reachable-but-never-reread heap: the live-heap
//! signature a profiler attributes to this allocation site.

/// Allocate `chunks` chunks of `chunk_bytes` each and retain them all. The
/// caller must hold the result alive while the heap profile is dumped.
pub fn grow(chunks: usize, chunk_bytes: usize) -> Vec<Vec<u8>> {
    let mut retained = Vec::with_capacity(chunks);
    for i in 0..chunks {
        retained.push(vec![(i % 251) as u8; chunk_bytes]);
    }
    retained
}

pub fn retained_bytes(retained: &[Vec<u8>]) -> usize {
    retained.iter().map(Vec::len).sum()
}

File: modules/java_workloads/src/RetainedGrowth.java
package monorepo.workloads;

import java.util.Arrays;

/**
 * Steadily grows a retained, reachable-but-never-reread heap: the
 * live-heap signature a profiler attributes to this allocation site.
 */
public final class RetainedGrowth {
    private RetainedGrowth() {
    }

    /**
     * Allocates chunks chunks of chunkBytes each and retains them all.
     */
    public static byte[][] grow(int chunks, int chunkBytes) {
        byte[][] retained = new byte[chunks][];
        for (int i = 0; i < chunks; i++) {
            byte[] chunk = new byte[chunkBytes];
            Arrays.fill(chunk, (byte) (i % 251));
            retained[i] = chunk;
        }
        return retained;
    }

    public static long retainedBytes(byte[][] retained) {
        long total = 0;
        for (byte[] chunk : retained) {
            total += chunk.length;
        }
        return total;
    }
}
